src/vma/proto/netlink_socket_mgr.h

```diff
@@ -289,20 +202,28 @@ void netlink_socket_mgr <Type>::update_tbl()
 //		len				: length of received data.
 //		p_ent_num		: number of rows in received data.
 template <typename Type>
-void netlink_socket_mgr <Type>::parse_tbl(int len, int *p_ent_num)
+void netlink_socket_mgr<Type>::parse_tbl_from_latest_cache(struct nl_cache *cache_state)
 {
-	struct nlmsghdr *nl_header;
-	int entry_cnt = 0;
+	uint16_t entry_cnt = 0;
 
-	nl_header = (struct nlmsghdr *) m_msg_buf;
-	for(;NLMSG_OK(nl_header, (u_int)len) && entry_cnt < MAX_TABLE_SIZE; nl_header = NLMSG_NEXT(nl_header, len))
+	struct nl_iterator_context
 	{
-		if (parse_enrty(nl_header, &m_tab.value[entry_cnt])) {
-			entry_cnt++;
+		Type * p_val_array;
+		uint16_t& entry_cnt;
+		netlink_socket_mgr<Type> * this_ptr;
+	} iterator_context = {m_tab.value, entry_cnt, this};
+
+	// a lambda can't be casted to a c-fptr with ref captures - so we provide context ourselves
+	nl_cache_foreach(cache_state, [](struct nl_object * nl_obj, void *context) {
+		nl_iterator_context* operation_context = reinterpret_cast<nl_iterator_context *>(context);
+		const bool is_valid_entry = operation_context->this_ptr->parse_entry(nl_obj, operation_context->p_val_array + operation_context->entry_cnt);
+		if (is_valid_entry)
+		{
+			++operation_context->entry_cnt;
 		}
-	}
-	if (p_ent_num)
-		*p_ent_num = entry_cnt;
+	}, &iterator_context);
+
+	m_tab.entries_num = entry_cnt;
```

Comment:
using this approach probably it is better insert sanity check inside this function
```
if (m_tab.entries_num >= MAX_TABLE_SIZE) {
		__log_warn("reached the maximum route table size");
	}
```

so that initialization and verification are performed in one place

Reply:
That's a good idea. fixed